x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/cti/threat_match_rows.tsx

```diff
@@ -5,44 +5,136 @@
-import React, { Fragment } from 'react';
+import type { FC } from 'react';
+import React, { Fragment, useState, useCallback } from 'react';
 import styled from 'styled-components';
 
+import type { EcsSecurityExtension } from '@kbn/securitysolution-ecs';
 import { ENRICHMENT_DESTINATION_PATH } from '../../../../../../../common/constants';
 import type { RowRenderer } from '../../../../../../../common/types';
 import type { Fields } from '../../../../../../../common/search_strategy';
 import { ID_FIELD_NAME } from '../../../../../../common/components/event_details/event_id';
 import { RowRendererContainer } from '../row_renderer';
 import { ThreatMatchRow } from './threat_match_row';
+import {
+  ALL_INDICATOR_MATCHES_MODAL_CLOSE,
+  ALL_INDICATOR_MATCHES_MODAL_HEADER,
+  SHOW_ALL_INDICATOR_MATCHES,
+} from '../translations';
 
 const SpacedContainer = styled.div`
   margin: ${({ theme }) => theme.eui.euiSizeS} 0;
 `;
 
-export const ThreatMatchRows: RowRenderer['renderRow'] = ({ data, isDraggable, scopeId }) => {
+export const renderThreatMatchRows: RowRenderer['renderRow'] = ({ data, isDraggable, scopeId }) => {
+  return <ThreatMatchRowWrapper data={data} isDraggable={isDraggable} scopeId={scopeId} />;
+};
+
+interface ThreatMatchRowProps {
+  data: EcsSecurityExtension;
+  isDraggable: boolean;
+  scopeId: string;
+}
+
+const MAX_INDICATOR_VISIBLE = 2;
+
+const ThreatMatchRowWrapper: FC<ThreatMatchRowProps> = ({ data, isDraggable, scopeId }) => {
   const indicators = get(data, ENRICHMENT_DESTINATION_PATH) as Fields[];
   const eventId = get(data, ID_FIELD_NAME);
 
+  const getThreatMatchRows = useCallback(
+    (mode: 'max' | 'all' = 'max') => {
+      const allIndicators =
+        mode === 'max' ? indicators.slice(0, MAX_INDICATOR_VISIBLE) : indicators;
+
+      return (
+        <RowRendererContainer data-test-subj="threat-match-row-renderer">
+          <SpacedContainer>
+            {allIndicators.map((indicator, index) => {
+              const contextId = `threat-match-row-${scopeId}-${eventId}-${index}`;
+              return (
+                <Fragment key={contextId}>
+                  <ThreatMatchRow
+                    contextId={contextId}
+                    data={indicator}
+                    eventId={eventId}
+                    isDraggable={isDraggable}
+                  />
+                  {index < indicators.length - 1 && <EuiHorizontalRule margin="s" />}
+                </Fragment>
+              );
+            })}
+          </SpacedContainer>
+        </RowRendererContainer>
+      );
+    },
+    [indicators, eventId, isDraggable, scopeId]
+  );
+
+  return (
+    <EuiFlexGroup direction="column" justifyContent="center" alignItems="center" gutterSize="none">
+      <EuiFlexItem>{getThreatMatchRows()}</EuiFlexItem>
+      {indicators.length > 10 && (
+        <EuiFlexItem>
+          <ThreatMatchRowModal title={SHOW_ALL_INDICATOR_MATCHES(indicators.length)}>
+            {getThreatMatchRows('all')}
```

Comment:
I wonder if it would make sense to move the modal logic into this component as well. Reason being that currently the react tree still contains all indicator matches. They're not rendered to the DOM since the modal is closed, but they still take up memory.

By moving the modal here, you can make sure the views are not part of the React tree when they're not needed.

Wdyt?